### ISSUE TITLE
Failure on to compile for armeabi using android NDK

### ISSUE THREAD

**Issue by cawka** [weidai11/cryptopp#119]
Here is the output of the compilation with gcc-4.9 or gcc-5 toolchains of Android NDK (official 10e and  CrystaX 10.3.1):

```
[armeabi] Compile++ thumb: cryptopp_static <= integer.cpp
/opt/android-ndk-r10e/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-g++ -MMD -MP -MF /Users/cawka/Devel/devel/cryptopp/extras/obj/local/armeabi/objs/cryptopp_static/__/__/integer.o.d -fpic -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -march=armv5te -mtune=xscale -msoft-float -fno-exceptions -fno-rtti -mthumb -Os -g -DNDEBUG -fomit-frame-pointer -fno-strict-aliasing -finline-limit=64 -I/opt/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include -I/opt/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/libs/armeabi/include -I/opt/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/backward -I/Users/cawka/Devel/devel/cryptopp/extras/jni -DANDROID  -Wa,--noexecstack -Wformat -Werror=format-security    -fexceptions -frtti  -I/opt/android-ndk-r10e/platforms/android-21/arch-arm/usr/include -c  /Users/cawka/Devel/devel/cryptopp/extras/jni/../../integer.cpp -o /Users/cawka/Devel/devel/cryptopp/extras/jni/../../integer.cpp: In constructor 'CryptoPP::InitializeInteger::InitializeInteger()':
/Users/cawka/Devel/devel/cryptopp/extras/jni/../../integer.cpp:2792:25: error: invalid conversion from 'bool (*)(const std::type_info&, void*, const void*)' to 'CryptoPP::PAssignIntToInteger {aka bool (*)(const std::type_info&, void*, const void*)}' [-fpermissive]
   g_pAssignIntToInteger = AssignIntToInteger;
                         ^
/Users/cawka/Devel/devel/cryptopp/extras/jni/../../integer.cpp: At global scope:
/Users/cawka/Devel/devel/cryptopp/extras/jni/../../integer.cpp:4491:10: warning: #pragma GCC target is not supported for this machine [-Wpragmas]
 # pragma GCC pop_options
          ^
[armeabi] Compile++ thumb: cryptopp_static <= md2.cpp
```

I'm kind of puzzled on what exactly compiler doesn't like: the suggested conversion seem to be exactly the same type.

Note that the issue doesn't exist for x86, x86_64, mips (and may be others) platforms.  If I compile with gcc 4.8, clang3.7 toolchain, the issue also doesn't exist.

Is it some kind of compiler (gcc 4.9, gcc 5.0) bug or there is some missing flag?

**Comment by cawka** [comment]
I have nailed down this issue to https://github.com/weidai11/cryptopp/issues/31 and commits 9fca0c28023a177106cf58a3de6da610f185a6e4 and dce2317195a7d9aa77b159fd1beddaf8358f6243 

I think there is some compiler optimization for the type declaration (with a non-hacked O flag) and the assignment.

**Comment by noloader** [comment]
Thanks Alex. Let me try and duplicate it.

> I have nailed down this issue to #31 and commits 9fca0c2 and dce2317

That's kind of a drag... It sounds like the proverbial *"pick your poison"*.

I reported the issue upstream, but I don't think Debian acted on it. Its kind of water under the bridge since none of the GCC bug fixes will show up in mainline. That is, it will stay broke in the distro like jessie/main, even if GCC fixes it. It might be backported and show up in jessie/updates, but its hard to say.

Does a ***`... && !defined(__ANDROID__)`*** help?

**Comment by noloader** [comment]
@Alex... the more I think about this....

I think we probably need a (3). That is, a `typedef` that sets up the type. I *suspect* its missing because of a circular dependency. That is, *`.integer.h`* depends upon *`algparam.h`* and vice versa. In this case, we may need to move the `typedef` of `PAssignIntToInteger` into *`cryptlib.h`* so its available to both `Integer` and `NameValuePairs`. We might need to forward declare some things in *`cryptlib.h`*.

**Comment by cawka** [comment]
(1) That was my first attempt to solve the problem.  Didn't help and the compiler is not helpful (just keep saying that the same types are not compatible)

(2) Adding !defined solved the issue

```patch
diff --git a/integer.cpp b/integer.cpp
index c532168..0ea8373 100644
--- a/integer.cpp
+++ b/integer.cpp
@@ -59,7 +59,7 @@
 #endif

 // Debian QEMU/ARMEL issue in MultiplyTop; see http://github.com/weidai11/cryptopp/issues/31.
-#if __ARMEL__ && (CRYPTOPP_GCC_VERSION >= 40900) && (CRYPTOPP_GCC_VERSION < 70000) && __OPTIMIZE__
+#if !defined(__ANDROID__) && __ARMEL__ && (CRYPTOPP_GCC_VERSION >= 40900) && (CRYPTOPP_GCC_VERSION < 70000) && __OPTIMIZE__
 # define WORKAROUND_ARMEL_BUG 1
 #endif
```

(3) I think (but haven't tried) that it would only work if type declaration, extern variable declaration, variable definition, and assignment all done with the same `-O1` hack.   Not sure how to achieve this...

**Comment by noloader** [comment]
Thanks Alex.

*"...and assignment all done with the same -O1 hack"* - that's an interesting observation. That could indicate we did no completely remediate the problems with GCC 4.9+ . Or it could be another one-off problem with Android's port of GCC.

Let me test (3) and see what happens. Whatever happens, this is your baby, and I hope you will provide the PR. I like it when folks make appearances at the [Contributors](http://github.com/weidai11/cryptopp/graphs/contributors) page. It reinforces the notion that the Community really does have an ownership stake.

**Comment by cawka** [comment]
(side note: you can always give attribution of some git commit to somebody else.  Just add `--author "Name <email>"` during `git commit`. I have exactly the same intention for my projects and it is sometimes faster this way.)

**Comment by noloader** [comment]
@cawka, Alex,

TLDR; ... this is bigger than just Android. It appears to affect 32-bit ARM and only surfaces with ***`-Os`***.

It looks like ***`... && !defined(__ANDROID__)`*** is an incomplete remediation. I think we need to go with the cast to ***`CryptoPP::PAssignIntToInteger`***.

*****

I checked this on a HiKey and BeagleBone. Both are real hardware and I use them to cross-validate results like this. The HiKey is ARM64 and uses Linaro with GCC 4.9.2-10. The BeagleBone is ARM32 and uses Debian with GCC 4.9.2-10.

The HiKey was fine, but the BeagleBone reported:

```
Testing started: Fri Jan 22 19:30:19 EST 2016
Testing finished: Sat Jan 23 10:25:31 EST 2016

24 configurations tested
8 errors detected
```

And:

````
4107:make: *** [libcryptopp.a] Error 1
4108:ERROR: failed to make cryptest.exe
21985:integer.cpp:2792:25: error: invalid conversion from ‘bool (*)(const std::type_info&, void*, const void*)’ to ‘CryptoPP::PAssignIntToInteger {aka bool (*)(const std::type_info&, void*, const void*)}’ [-fpermissive]
21993:make: *** [integer.o] Error 1
21994:ERROR: failed to make cryptest.exe
22113:integer.cpp:2792:25: error: invalid conversion from ‘bool (*)(const std::type_info&, void*, const void*)’ to ‘CryptoPP::PAssignIntToInteger {aka bool (*)(const std::type_info&, void*, const void*)}’ [-fpermissive]
22121:make: *** [integer.o] Error 1
22122:ERROR: failed to make cryptest.exe
24975:8 errors detected
```

So, out of 24 builds, 2 of them failed (and caused the script to detect 8 errors). Probing deeper on the BeagleBone, this only occurs with ***`-Os`***:

```
g++ -DDEBUG -g2 -Os  -Wno-deprecated-declarations -fPIC -pipe -c integer.cpp
integer.cpp: In constructor ‘CryptoPP::InitializeInteger::InitializeInteger()’:
integer.cpp:2792:25: error: invalid conversion from ‘bool (*)(const std::type_info&, void*, const void*)’ to ‘CryptoPP::PAssignIntToInteger {aka bool (*)(const std::type_info&, void*, const void*)}’ [-fpermissive]
   g_pAssignIntToInteger = AssignIntToInteger;
```

We can duplicate easily with:

```
$ make CXXFLAGS="-DDEBUG -g2 -Os  -Wno-deprecated-declarations -fPIC -pipe" integer.o
g++ -DDEBUG -g2 -Os  -Wno-deprecated-declarations -fPIC -pipe -c integer.cpp
integer.cpp: In constructor ‘CryptoPP::InitializeInteger::InitializeInteger()’:
integer.cpp:2792:25: error: invalid conversion from ‘bool (*)(const std::type_info&, void*, const void*)’ to ‘CryptoPP::PAssignIntToInteger {aka bool (*)(const std::type_info&, void*, const void*)}’ [-fpermissive]
   g_pAssignIntToInteger = AssignIntToInteger;
```

After performing the cast, the issue clears itself (modulo the warning). Can you test the following and make a PR after verifying it:

```
$ git diff integer.cpp 
diff --git a/integer.cpp b/integer.cpp
index c532168..dac50b9 100644
--- a/integer.cpp
+++ b/integer.cpp
@@ -2789,7 +2789,7 @@ InitializeInteger::InitializeInteger()
        if (!g_pAssignIntToInteger)
        {
                SetFunctionPointers();
-               g_pAssignIntToInteger = AssignIntToInteger;
+               g_pAssignIntToInteger = (CryptoPP::PAssignIntToInteger)AssignIntToInteger;
        }
 }
```

**Comment by cawka** [comment]
Interesting.  I have checked with the cast (`static_cast`), but forgot to check with C-style cast.  With this C-style cast (commit 96c51ffb2c5ea8e8f419193325fdd69c61199712), compilation worked.

I have also checked the validation tests (`cryptest.exe v` and `cryptest.exe tv all`) on android armeabi emulator and there were no failures.

**Comment by cawka** [comment]
(You can use the following command to merge the commit: `git fetch https://github.com/cawka/cryptopp-1 master && git merge FETCH_HEAD`)

**Comment by noloader** [comment]
> You can use the following command to merge the commit: git fetch https://github.com/cawka/cryptopp-1 master && git merge FETCH_HEAD

OK, thanks Alex. So I performed:

```
# My cryptopp directory
cd cryptopp

# Back to the original code
git reset --hard HEAD

# Pull Alex changes
git fetch https://github.com/cawka/cryptopp-1 master && git merge FETCH_HEAD
From https://github.com/cawka/cryptopp-1
 * branch            master     -> FETCH_HEAD
Already up-to-date.

# Now check there's something to check-in:
git diff
```

Git is telling me there is nothing to check-in. I also checked [Commit History](http://github.com/weidai11/cryptopp/commits/master), and I don't see your fix checked-in.

I don't know how to commit something that Git is not aware of. But I also know Git hides some differences, like when a tag is added, so I have to take special steps to perform the check-in.

Would you happen to know what I am missing?

**Comment by noloader** [comment]
@cawka, Alex,

Cancel that request. I performed a ***`git push`*** and the non-existent changes were checked-in.

Gotta love Git. I tries to make even simple tasks difficult...

**Comment by cawka** [comment]
That's strange.  The command I gave you should have worked (I use something like this all the time), unless you somehow already had this code checked out...

```bash
$ git checkout -B tmp df1f68da841c4be9213a7637a3a2abb0b5edd803
Reset branch 'tmp'

$ git fetch https://github.com/cawka/cryptopp-1 master && git merge FETCH_HEAD
From https://github.com/cawka/cryptopp-1
 * branch            master     -> FETCH_HEAD
Updating df1f68d..96c51ff
Fast-forward
 integer.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

$ git diff HEAD^
diff --git a/integer.cpp b/integer.cpp
index c532168..dac50b9 100644
--- a/integer.cpp
+++ b/integer.cpp
@@ -2789,7 +2789,7 @@ InitializeInteger::InitializeInteger()
        if (!g_pAssignIntToInteger)
        {
                SetFunctionPointers();
-           g_pAssignIntToInteger = AssignIntToInteger;
+         g_pAssignIntToInteger = (CryptoPP::PAssignIntToInteger)AssignIntToInteger;
        }
 }
```
